### PR TITLE
Fixed the issue with Dial-based JMM2

### PR DESCRIPTION
**Pull request by mar1akc**
There used to be a logical error in switching between getting perpendicular of (x,y) using (-y,x) or (y,-x). This matters in my implementation only for JMM2. Just because points are accepted in different order in Dial and Dijkstra templates, this issue did not affect the Dijkstra-based JMM2 by luck.

### CODE REVIEW

**Comment by mar1akc** [comment]
I made a number of changes in my package. Now the only global variables are: slo_fun, method_template, and method_update.

**Comment by mar1akc** [comment]
The main code now is JMMversion3.c

File: BucketSort.c [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+void dial_bucket_init(struct mybucket *bucket,int iskip,double gap);
+void print_buckets(int Nbuckets,struct mybucket *bucket);
+int adjust_bucket(int ind,double newval,double g,int Nbuckets,struct mybucket *bucket,struct mylist *list);
+int find_bucket(double utemp,double g);
+void myfree(struct bucket_sort_stuff  *BB);
+void start_filling_buckets(struct bucket_sort_stuff  *BB,int Nbuckets,struct mybucket *bucket,
+		struct mylist *list,double gap,int *bdry,double *blist,int bcount);
+int find_number_of_buckets(double gap,double maxgap);
+void form_list_of_new_valid_points(struct mybucket *bucket,int *newlist,int *empty_count);
+//---------------------------------------------------------------
```
Comment (sampotter) [inline BucketSort.c:15]: Please remove this kind of "//------" comment.

File: BucketSort.c [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+
+void dial_list_init(struct mylist *list,int ind) {
+	list -> ind = ind;
+	list -> previous = NULL;
+	list -> next = NULL;
+	list -> ibucket = -1; // no bucket is assigned
```
Comment (sampotter) [inline BucketSort.c]: Replace `-1` with `NO_INDEX` from def.h.

File: BucketSort.c [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+#include "QuickSort.h"
+
+#define INFTY 1.0e+6
+#define TOL 1.0e-14
+#define mabs(a) ((a) >= 0 ? (a) : -(a))
+#define sgn(a) ((a) == 0 ? 0 : ((a) > 0  ? 1 : -1 ))
+#define max(a,b) ((a) >= (b) ? (a) : (b))
+#define min(a,b) ((a) <= (b) ? (a) : (b))
+
+void dial_list_init(struct mylist *list,int ind);
```
Comment (sampotter) [inline BucketSort.c]: You don't need to forward declare these functions, since they're already declared in "BucketSort.h".

**Comment by sampotter** [comment]
There are several different topics for this PR (pull request). Usually there should just be one.

When there's a single topic for a PR it makes it easier to use the discussion and comments to focus on this PR. You can create as many PRs simultaneously as you want to separate things into different topics. Since a PR corresponds to "merging the contents of one branch into another branch", it then becomes important to assign separate branchs to separate topics. 

The two main topics from what I can tell are:

- Fixing a bug with JMM2 (the original title of the PR)
- Some more [refactoring](https://en.wikipedia.org/wiki/Code_refactoring) ("modularization") where you separate out a few structs into different modules (e.g. your bucket sort stuff and a mesh struct)

So far I have had a chance to look at the bucket refactoring. I left a couple of comments on specific lines of the files that you added/changed. Please let me know if you can see them. I think you should be able to reply directly to those comments or fix them.

Additionally, I'll make a few comments about the refactoring. I'm mostly making this suggestion here before accepting your PR so that we can practice this style of developing. These comments are just related to the style/organization and not the logical correctness of your code.

You have `struct bucket_sort_stuff` and `mybucket` and `mylist`. Some remarks and suggestions:

- Names should never be diminutive or informal, as in "mysomething" or "something_stuff".
- Names should strike a balance between descriptiveness and brevity (i.e. they should be "parsimonious"). For example, `mylist` is short but does not inform the user what it's purpose is. From what I remember, `mylist` is the bucket sort version of the heap back pointer required by Dial's algorithm. A better name might be "bucket_backptr".
- Another point about `mylist`: if you were to use the opaque struct technique and move the "implementation details" of your bucket sort into BucketSort.c, then it would be clear from context that `bucket_backptr` is a back pointer for bucket sort, in which case `backptr` would be clear enough.
- Your bucket sort has the following structure: there is a "main" struct (`bucket_sort_stuff`) which controls the subsidiary structs (`mylist` and `mybucket`). In an implementation of a bucket sort with the right interface, only `bucket_sort_stuff` should be exposed in "BucketSort.h" (and it should have a more descriptive name without "stuff" in it). The other structs should only exist in "BucketSort.c" and be invisible outside of this file. If you can't easily make this change, you should ask why and determine if the dependency preventing this is real or imagined.
- There are many functions which duplicate some information already available in `bucket_sort_stuff`, in particular `Nbuckets`. This variable is already in `bucket_sort_stuff`, so it's wasteful/confusing/unnecessary to pass it in two different places.

I am not trying to be mean with my comments, just helpful. :-) Also, if any of this is confusing, PLEASE ask, I am happy to help via Zoom. I'm also sort of figuring out how to do some of this stuff myself.

**Comment by mar1akc** [comment]
We have just checked in to our cabin in Bethany Beach, Delaware.
Thanks for all your comments—I see them in emails now. I hope to be able to
work tonight and see them on github. I will address your comments then.

I did lots of programming last weekend and I have a lot of plans for
working on my code. I cannot start a new PR until you accept the previous
one—github won’t let me. That’s why there were  several topics.
My only options are to cancel my request or stop submitting commits:)

After I am happy with my parallelizable version of Dial-based JMM, I want
to start replacing my packages with yours. I believe I can replace linear
algebra  and adopt def.h on my own. Then we can try to insert your heap
stuff instead of mine.

**Comment by sampotter** [comment]
> I cannot start a new PR until you accept the previous one—github won’t let me. That’s why there were several topics.

Hmm, weird! I didn't realize it was so restrictive. Let me look into this... If this is really the case I'll have to be faster with accepting your PRs.

I have more comments on your code---I think it would be good PR practice for me to finish with my comments and have you address them before I accept this PR.

You can continue to make commits on a separate branch while we finish with this PR, and they won't affect this one. Git can "merge" two branches, so if you make a new branch off of develop and continue working on both develop and your new branch a bit in parallel, so long as they don't completely clobber each other, git should be able to figure things out (you can also manually help it resolve a conflicted merge).

Please enjoy your vacation!

File: BucketSort.c (hunk 2 above)
Comment (mar1akc) [reply BucketSort.c]: Done

File: BucketSort.c (hunk 3 above)
Comment (mar1akc) [reply BucketSort.c]: Removed such redundant declaration in BucketSort.c and all other files with .hcounterpart.

File: BucketSort.c (hunk 1 above)
Comment (mar1akc) [reply BucketSort.c:15]: If you do want these separators gone I'll remove them but later. These separators help me find functions in the library files faster, and I am still editing them.

**Comment by mar1akc** [comment]
(1) I renamed: bucket_sort_stuff --> bucket_sort_handle, mylist --> backprt_list. Just "backprt" would not reflect the fact that I form link lists out of mesh points. Similarly:  binary_tree_stuff --> binary_tree_handle.
(2) I can pass around bucket_sort_handle instead of its components, but I find it more convenient to work with its components while within dial_main_body, especially while I am developing some new logic. Hence: (A) I can show you in my code why I prefer to pass around the components while inside dial_main_body, or (B) if you want to pass only bucket_sort_handle no matter what, I'll switch to it after I am done developing JMMversion4.c (parallelizable/vectorizable version of dial).

File: BucketSort.c (hunk 1 above)
Comment (sampotter) [reply BucketSort.c:15]: OK, no problem.

**Pull request merged**